### PR TITLE
Fix S1104 FP: Do not report in Unity serializable classes

### LINKED ISSUE #7522: Fix S1104 FP: Do not report in Unity serializable classes
Derived from [this discuss post](https://discuss.sonarsource.com/t/is-s1104-compatible-with-unity/14333/1). 

[S1104](https://sonarsource.github.io/rspec/#/rspec/S1104/csharp) raises in all contexts, including in Unity scripts derived from [MonoBehavior](https://docs.unity3d.com/ScriptReference/MonoBehaviour.html). The rule should not report in that context, however, due to how Unity works. 
Using public properties would break core features of Unity such as YAML serialization and UI editing.

Unity requires using public fields, and using public properties would break core features of Unity such as serialization and UI editing.

Unity has specific rules for serialization, detailed [here](https://docs.unity3d.com/Manual/script-Serialization.html).

> Serializers in Unity work directly on the fields of your C# classes rather than their properties, so there are rules that your fields must conform to be serialized.

Notice that there are ways of conforming to S1104 in Unity: as shown in the “Serialization of properties” section of the documentation, user-implemented properties with a backing field marked as SerializeField would be supported by the Unity Editor.

Regardless, those are corner cases that are rarely encountered in practice.

### PR DESCRIPTION
**Pull request by sebastien-marichal**
Fixes #7522 

Ignore fields in the following Unity classes:
- [ScriptableObject](https://docs.unity3d.com/ScriptReference/ScriptableObject.html)
- [MonoBehaviour](https://docs.unity3d.com/ScriptReference/MonoBehaviour.html)

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/FieldsShouldBeEncapsulatedInProperties.cs [hunk 1]
```diff
@@ -37,6 +37,10 @@ public sealed class FieldsShouldBeEncapsulatedInProperties : SonarDiagnosticAnal
         SyntaxKind.ConstKeyword
     };
 
+    private static readonly ImmutableArray<KnownType> IgnoredTypes = ImmutableArray.Create(
+        KnownType.UnityEngine_MonoBehaviour,
+        KnownType.UnityEngine_ScriptableObject);
```
Comment (mary-georgiou-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/FieldsShouldBeEncapsulatedInProperties.cs:42]: My understanding from the docs is that you should include `UnityEngine.Object`.
Which, if I'm not wrong, is parent to `UnityEngine_ScriptableObject` so you can replace one with the other.

Comment (sebastien-marichal) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/FieldsShouldBeEncapsulatedInProperties.cs:42]: Adding `UnityEngine.Object` is not enough, the class should also have the `Serializable` attribute.
I can add it so it also supports custom serializable unity class/object.
I will rework a bit the logic then.

Comment (sebastien-marichal) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/FieldsShouldBeEncapsulatedInProperties.cs:42]: After a closer look, the required attribute for custom objects is `System.Serializable`, and this is already an exception to the rule.

**Review by mary-georgiou-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**